### PR TITLE
Add driver for Bosch BMP180 temp/pressure sensor

### LINKED ISSUE #130: Driver for Bosch BMP180 (i2c temperature+pressure sensor)
Hi,

thanks a lot for this project! I've just switched over from https://github.com/kidoman/embd and like it very much. I have a couple of [BMP180](https://ae-bst.resource.bosch.com/media/_tech/media/datasheets/BST-BMP180-DS000-121.pdf) temperature and pressure sensors lying around, and I'm just implementing a driver for periph. I'll do a PR later this weekend.

### PR DESCRIPTION
**Pull request by fd0**
This PR adds a driver for the Bosch BMP180 temperature/pressure sensor. At the moment the driver works and reads sensible data.

Closes #130

### CODE REVIEW

File: devices/bmp180/bmp180.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
+// Copyright 2016 The Periph Authors. All rights reserved.
```
Comment (quinte17) [inline devices/bmp180/bmp180.go]: is here 2017 preferable?

File: cmd/bmp180/main.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
+// Copyright 2016 The Periph Authors. All rights reserved.
```
Comment (quinte17) [inline cmd/bmp180/main.go]: is here 2017 preferable?

File: devices/bmp180/bmp180.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+	d.c.AC3 = int16(cal[4])<<8 | int16(cal[5])
+	d.c.AC4 = uint16(cal[6])<<8 | uint16(cal[7])
+	d.c.AC5 = uint16(cal[8])<<8 | uint16(cal[9])
+	d.c.AC6 = uint16(cal[10])<<8 | uint16(cal[11])
+	d.c.B1 = int16(cal[12])<<8 | int16(cal[13])
+	d.c.B2 = int16(cal[14])<<8 | int16(cal[15])
+	d.c.MB = int16(cal[16])<<8 | int16(cal[17])
+	d.c.MC = int16(cal[18])<<8 | int16(cal[19])
+	d.c.MD = int16(cal[20])<<8 | int16(cal[21])
+
```
Comment (quinte17) [inline devices/bmp180/bmp180.go]: please consider something like
`convert(cal[:], &d.c)`
where
```
import "encoding/binary"
import "bytes"

func convert(b []byte, data interface{}) error {
	buf := bytes.NewReader(b)
	return binary.Read(buf, binary.LittleEndian, data)
}
```

File: cmd/bmp180/main.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// bmp180 reads the current temperature and pressure from a BMP180.
```
Comment (maruel) [inline cmd/bmp180/main.go]: /me start thinking about more coherency for package docstring
https://godoc.org/periph.io/x/periph/devices#pkg-subdirectories

At the very least, it shall start with: `Package bmp180 ...`

Include a Reference section with a link to the datasheet.
https://ae-bst.resource.bosch.com/media/_tech/media/datasheets/BST-BMP180-DS000-121.pdf

File: devices/bmp180/bmp180.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	env.Pressure = devices.KPascal(pressure)
+	return nil
+}
+
+// Halt is a noop for the BMP180.
+func (d *Dev) Halt() error {
+	return nil
+}
+
+// Opts is optional options to pass to the constructor.
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: Since there's only one parameter, pass the parameter to New() directly. No need for a struct.

File: devices/bmp180/bmp180.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+	regSoftReset        = 0xE0 // soft reset register
+	regCtrlMeas         = 0xF4 // control measurement
+	regOutMSB           = 0xF6 // MSB of measurement
+	regOutLSB           = 0xF7 // LSB of measurement
+	regOutXLSB          = 0xF8 // extended LSB of measurement
+)
+
+const softResetValue = 0xB6
+
+func (d *Dev) makeDev(opts *Opts) error {
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: Since there's no SPI support (the datasheet says it's on custom orders only, so not worth implementing), then there's no need of a makeDev() function.

File: cmd/bmp180/main.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	for {
+		var env devices.Environment
+		if err := e.Sense(&env); err != nil {
+			return err
+		}
+		fmt.Printf("%8s %10s %9s\n", env.Temperature, env.Pressure, env.Humidity)
+		if !loop {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
```
Comment (maruel) [inline cmd/bmp180/main.go]: Use the conversion time from the datasheet at page 21. 100ms is waaay too long.

File: devices/bmp180/bmp180.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+	Pressure Oversampling
+}
+
+const (
+	bmp180Address = 0x77 // the address of a BMP180 is fixed at 0x77
+	chipID        = 0x55 // contents of the ID register, always 0x55
+)
+
+// New returns an object that communicates over I²C to BMP180 environmental
+// sensor.
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: Add a note that the device can work at up to 3.4MHz. This is valuable information.

File: devices/bmp180/bmp180.go [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+	return nil
+}
+
+// Opts is optional options to pass to the constructor.
+type Opts struct {
+	Pressure Oversampling
+}
+
+const (
+	bmp180Address = 0x77 // the address of a BMP180 is fixed at 0x77
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: I'm not a huge fan of constants that are used exactly once.

File: devices/bmp180/bmp180_test.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
+package bmp180
```
Comment (maruel) [inline devices/bmp180/bmp180_test.go]: add standard header.

File: devices/bmp180/bmp180.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+	time.Sleep(100 * time.Millisecond)
+
+	// read value
+	var tempBuf [2]byte
+	if err := d.readReg(regOutMSB, tempBuf[:]); err != nil {
+		return err
+	}
+
+	UT := uint16(tempBuf[0])<<8 | uint16(tempBuf[1])
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: why uppercase?

File: devices/bmp180/bmp180.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+	// start conversion for pressure
+	cmd = 1 << ctrlStartConversion
+	cmd |= ctrlCommandPressure << ctrlMeasurementControl
+	cmd |= byte(d.os) << ctrlOversampling
+
+	if err := d.d.Tx([]byte{regCtrlMeas, cmd}, nil); err != nil {
+		return err
+	}
+
+	time.Sleep(100 * time.Millisecond)
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: too long

File: devices/bmp180/bmp180.go [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+	d.c.B1 = int16(cal[12])<<8 | int16(cal[13])
+	d.c.B2 = int16(cal[14])<<8 | int16(cal[15])
+	d.c.MB = int16(cal[16])<<8 | int16(cal[17])
+	d.c.MC = int16(cal[18])<<8 | int16(cal[19])
+	d.c.MD = int16(cal[20])<<8 | int16(cal[21])
+
+	return nil
+}
+
+func (d *Dev) readReg(reg uint8, b []byte) error {
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: You may want to use periph.io/x/periph/conn/mmr which exists exactly for this.

The bme280 driver couldn't use it because of the SPI weirdo communication protocol.

File: devices/bmp180/bmp180.go [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+	return d.d.Tx([]byte{reg}, b)
+}
+
+// calibration data read from the internal EEPROM (datasheet page 13)
+type calibration struct {
+	AC1, AC2, AC3 int16
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: why export the members? Please use lowercase even if they are uppercase in the datasheet.

**Comment by maruel** [comment]
gohci

**Comment by maruel** [comment]
(I need to fix the gohci check)

**Comment by maruel** [comment]
gohci: test

(sorry I'm using your PR as a test bed to enable gohci jobs on PRs)

File: devices/bmp180/bmp180.go (hunk 1 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Sure.

File: devices/bmp180/bmp180.go (hunk 3 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Hm, I disagree here:
 * Is that really necessary? I've use the same style as in the `bme280` driver, which uses exactly this here: https://github.com/google/periph/blob/master/devices/bme280/bme280.go#L257-L269
 *  It feels wrong to me. What's happening here is that the bytes are interpreted as big endian (not little endian) according to the datasheet. Having a separate function call, that allocates a bytes.Reader struct, only to read two bytes out of it feels like A LOT overhead.
 * IMHO it makes the code much harder to read.

File: cmd/bmp180/main.go (hunk 2 above)
Comment (fd0) [inline cmd/bmp180/main.go]: I think so.

File: cmd/bmp180/main.go (hunk 4 above)
Comment (fd0) [inline cmd/bmp180/main.go]: Uhm, that's not package `bmp180`, but package `main`. The package `bmp180` has a proper doc string and a link to the datasheet:
```
// Package bmp180 controls a Bosch BMP180 device over I²C.
//
// Datasheet
//
// https://cdn-shop.adafruit.com/datasheets/BST-BMP180-DS000-09.pdf
```

The datasheet you linked is really bad btw, the font in the algorithm on page 15 is unreadable. The one at adafruit that I linked has at least a readable font in the screenshot...

File: devices/bmp180/bmp180.go [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+
+// Possible oversampling values.
+const (
+	No  Oversampling = 0
+	O2x Oversampling = 1
+	O4x Oversampling = 2
+	O8x Oversampling = 3
+)
+
+// Dev is an handle to a bmp180.
```
Comment (tve) [inline devices/bmp180/bmp180.go]: s/an/a/

File: devices/bmp180/bmp180.go (hunk 11 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: I used the syntax from the manual.

File: cmd/bmp180/main.go (hunk 7 above)
Comment (fd0) [inline cmd/bmp180/main.go]: Ok

File: devices/bmp180/bmp180.go (hunk 12 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Agreed.

File: devices/bmp180/bmp180.go [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+
+// bit offsets in regCtrlMeas
+const (
+	ctrlMeasurementControl = 0
+	ctrlStartConversion    = 5
+	ctrlOversampling       = 6
+)
+
+// Sense returns measurements as °C and kPa.
+func (d *Dev) Sense(env *devices.Environment) error {
```
Comment (tve) [inline devices/bmp180/bmp180.go]: @maruel why doesn't sense return devices.Environment?

File: devices/bmp180/bmp180.go (hunk 5 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Ok.

File: cmd/bmp180/main.go (hunk 7 above)
Comment (tve) [inline cmd/bmp180/main.go]: Since this is in a loop printing to stdout maybe it should be something slower that one can follow? Such as 2x per second?

File: devices/bmp180/bmp180.go (hunk 9 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: I disagree: IMHO assigning a name helps readability a lot. This gives a magic constant that is suddenly used in the code a name, which can e.g. be searched in the datasheet. This is a common practice in development for embedded systems (at least in my experience). Thoughts?

File: devices/bmp180/bmp180.go (hunk 8 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: OK.

File: devices/bmp180/bmp180.go (hunk 6 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Ok.

File: devices/bmp180/bmp180.go (hunk 3 above)
Comment (tve) [inline devices/bmp180/bmp180.go]: To me the `convert` idea is pretty slick. It could actually be a method on `calibration` that takes the byte array as arg and returns a `calibration` struct, mhh, could actually be newCalibration. You're creating a binary.BigEndian reader for the entire struct, not for each 2-byte element. Or am I not getting it right?

Comment (maruel) [inline devices/bmp180/bmp180.go]: bme280 uses custom because of weirdo bit packing at nibble boundary for the "h" ones.

https://periph.io/x/periph/conn/mmr#Dev8.ReadStruct is what you want.

File: devices/bmp180/bmp180.go (hunk 9 above)
Comment (maruel) [inline devices/bmp180/bmp180.go]: You are in package bmp180, so prefixing a const with "bmp180" is redundant.

It was more of a nit than anything else, I don't mind much.

File: devices/bmp180/bmp180.go (hunk 16 above)
Comment (maruel) [inline devices/bmp180/bmp180.go]: in case the struct grows a lot to save on allocation, I agree it could be changed, I'm not confident with this interface.

File: cmd/bmp180/main.go (hunk 7 above)
Comment (maruel) [inline cmd/bmp180/main.go]: Oh right, for printing to the user, you can make this configurable through https://golang.org/pkg/flag/#Duration with `time.Second` the default. Also you want once per second, not sleep one second, which means using https://golang.org/pkg/time/#After

(I haven't done thing in bme280/main.go, it'd be a good thing to fix there too)

File: cmd/bmp180/main.go (hunk 4 above)
Comment (maruel) [inline cmd/bmp180/main.go]: Oops sorry I was really confused, ignore my comment about package name.

It kind of suck that the official one is of less quality. I'm not a fan of linking to adafruit's copies.

File: devices/bmp180/bmp180.go (hunk 11 above)
Comment (maruel) [inline devices/bmp180/bmp180.go]: still, upper case has special meaning in Go so use lower case.

File: devices/bmp180/bmp180.go [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+
+const (
+	ctrlCommandTemperature = 0x0E
+	ctrlCommandPressure    = 0x14
+)
+
+// bit offsets in regCtrlMeas
+const (
+	ctrlMeasurementControl = 0
+	ctrlStartConversion    = 5
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: These are bitshift, so their name should be ctrlMeasurementShift, ctrlStartConversionShift and ctrlOversampleShift. Otherwise the name is confusing.

File: devices/bmp180/bmp180.go [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+	c  calibration
+	os Oversampling
+}
+
+func (d *Dev) String() string {
+	return fmt.Sprintf("BMP180{%s}", d.d)
+}
+
+const (
+	ctrlCommandTemperature = 0x0E
```
Comment (maruel) [inline devices/bmp180/bmp180.go]: Frankly it'd make more sense if the constants were fixed here, e.g.
```
ctrlCommandTemperature = 0x0E << ctrlMeasurementControl
```

but since these commands cannot run without the start conversion bit, I don't see value in having these constant separated.

File: devices/bmp180/bmp180.go (hunk 16 above)
Comment (tve) [inline devices/bmp180/bmp180.go]: given everything else, the allocation of the struct seems irrelevant to me...

File: devices/bmp180/bmp180.go (hunk 3 above)
Comment (quinte17) [inline devices/bmp180/bmp180.go]: the convert idea gets all your single lines into a oneliner..
ok maybe i missed the endianess.
example usage can be seen here:
https://github.com/quinte17/bme280/blob/master/bme280.go#L357
but i agree for bme280 its not perfect either..

Comment (fd0) [inline devices/bmp180/bmp180.go]: @quinte17 Ah! I re-read your comment and I missed that you pass a pointer to the calibration struct to `convert()`, now it makes sense. Thanks!

File: devices/bmp180/bmp180.go (hunk 11 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Hm, I'm not aware of any special meaning of the case of variables within a function, am I missing something?

I'll use lower case.

Comment (maruel) [inline devices/bmp180/bmp180.go]: You are right there is no meaning inside a function, but the code style is to always use lower case names for locally scoped symbols.

File: cmd/bmp180/main.go (hunk 4 above)
Comment (fd0) [inline cmd/bmp180/main.go]: I agree. Maybe we can put both links in? I also found at least two errors in the official datasheet...

Comment (maruel) [inline cmd/bmp180/main.go]: Sure!

BTW sorry for the feeling of nitpicking fest; I'm trying to keep a very specific style on the library and for the device drivers I really want to behave the best w.r.t. the device (e.g. like sleeping the exact time as specified by the datasheet). I hope it's not a bummer for you.

File: cmd/bmp180/main.go (hunk 7 above)
Comment (fd0) [inline cmd/bmp180/main.go]: Fine with me.

File: devices/bmp180/bmp180.go (hunk 9 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Ok, I can live with that :)

What I meant was mainly tailored towards magic constants that have been sourced from somewhere else (e.g. the datasheet).

File: devices/bmp180/bmp180.go (hunk 13 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Ah, thanks! I wasn't aware of the mmr package.

File: devices/bmp180/bmp180.go (hunk 14 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Hm, I'm pretty sure nothing is exported here since the struct itself is not exported...

Comment (fd0) [inline devices/bmp180/bmp180.go]: That doesn't work: Using reflection on not exported struct members fails, so we cannot use `binary.Read()` to parse the calibration data. I'm leaving this upper case.

Besides: The struct itself is not exported, so it's not as if it will show up somewhere, not even in godoc...

File: devices/bmp180/bmp180_test.go (hunk 10 above)
Comment (fd0) [inline devices/bmp180/bmp180_test.go]: "header" as in "package header" or "copyright information"?

File: devices/bmp180/bmp180.go (hunk 14 above)
Comment (maruel) [inline devices/bmp180/bmp180.go]: Oh sure, if (and only if) you switch to using reflection, then keep them exported.

The idea here is to always aim for least privilege. Uppercase/lowercase is a way to document to the _reader of the code_ if there are interactions that falls outside of the assignments as shown in the package.

File: devices/bmp180/bmp180_test.go (hunk 10 above)
Comment (maruel) [inline devices/bmp180/bmp180_test.go]: Copyright header.

File: devices/bmp180/bmp180.go (hunk 16 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: I agree, it's irrelevant. But I also think it is outside the scope of this PR to change the API :)

File: devices/bmp180/bmp180.go (hunk 17 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: Agreed.

File: devices/bmp180/bmp180.go (hunk 18 above)
Comment (fd0) [inline devices/bmp180/bmp180.go]: I've renamed the constants here to `cmdStartTempConv`, I think that makes more sense. Thanks for the suggestion.

**Comment by fd0** [comment]
I've addressed all comments and moved the device to `experimental/devices` (and I've verified that the driver still works of course). Looking forward to further comments!

File: cmd/bmp180/main.go [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+		}
+
+		<-t.C
+	}
+	return nil
+}
+
+func mainImpl() error {
+	i2cID := flag.String("i2c", "", "I²C bus to use")
+	hz := flag.Int("hz", 0, "I²C bus speed")
```
Comment (maruel) [inline cmd/bmp180/main.go]: I'd say to not bother, I had added these because I was experimenting but it's really irrelevant for a sensing device, I'll post a change to remove it from the bme280 exec.

Also setspeed is not working fine on the rpi so it's not even usable at the time. :(

File: cmd/bmp180/main.go [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+	sample2x := flag.Bool("s2", false, "sample at 2x")
+	sample4x := flag.Bool("s4", false, "sample at 4x")
+	sample8x := flag.Bool("s8", false, "sample at 8x")
+	loop := flag.Bool("l", false, "read data continously")
+	interval := flag.Duration("i", time.Second, "read data continously with this interval")
```
Comment (maruel) [inline cmd/bmp180/main.go]: what about just having -i, make it default to 0?
then no need for a second flag. will update bme280 later.

File: cmd/bmp180/main.go [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	if *hz != 0 {
+		if err := bus.SetSpeed(int64(*hz)); err != nil {
+			return err
+		}
+	}
+
+	var dev *bmp180.Dev
+
+	fmt.Printf("open bmp180\n")
```
Comment (maruel) [inline cmd/bmp180/main.go]: don't forget to remove

File: cmd/bmp180/main.go [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+	if dev, err = bmp180.New(bus, os); err != nil {
```
Comment (maruel) [inline cmd/bmp180/main.go]: extract the assignment from the if, so no need to define dev upfront

File: experimental/devices/bmp180/bmp180.go [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+	// issue soft reset to initialize device
+	if err := d.dev.WriteUint8(regSoftReset, softResetValue); err != nil {
+		return nil, err
+	}
+
+	// wait for restart
+	time.Sleep(10 * time.Millisecond)
+
+	// read calibration data from internal EEPROM, 11 registers with two bytes each
+	var cal [22]byte
```
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: cal := calibration {}

File: experimental/devices/bmp180/bmp180.go [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+	env.Pressure = devices.KPascal(pressure)
+	return nil
+}
+
+// Halt is a noop for the BMP180.
+func (d *Dev) Halt() error {
+	return nil
+}
+
+const (
```
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: group all consts groups without functions in between.

but: I generally prefer public symbols to be first, the. private .
then each group has const, var, the code. it's not an hard requirement

File: experimental/devices/bmp180/bmp180.go (hunk 23 above)
Comment (fd0) [inline experimental/devices/bmp180/bmp180.go]: But then we cannot do the check a few lines later easily, as the datasheet suggests...

Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: The main reason I prefer this is because of the current code use reflection twice. It is mostly a bug in the API design of `mmr` but still. The fact that big endian decoding is done part manually, part automatically, is kind of weird.

```
func isValid(i int16) bool {
  return i != 0 && i != ^int16(0)
}
func isValidU(i uint16) bool {
  return i != 0 && i != 0xFFFF
}

func (c*calibration) isValid() bool {
  return isValid(c.AC1) && isValid(c.AC2) && isValid(c.AC3) && isValidU(c.AC4) && isValidU(c.AC5) && isValidU(c.AC6) && isValid(c.B1) && isValid(c.B2) && isValid(c.MB) && isValid(c.MC) && isValid(c.MD)
}
```

Comment (quinte17) [inline experimental/devices/bmp180/bmp180.go]: should the newCalibration func validate before returning a correct instance?

File: cmd/bmp180/main.go (hunk 19 above)
Comment (fd0) [inline cmd/bmp180/main.go]: OK

File: cmd/bmp180/main.go (hunk 20 above)
Comment (fd0) [inline cmd/bmp180/main.go]: Good idea.

File: cmd/bmp180/main.go (hunk 22 above)
Comment (fd0) [inline cmd/bmp180/main.go]: OK

File: experimental/devices/bmp180/bmp180.go (hunk 23 above)
Comment (fd0) [inline experimental/devices/bmp180/bmp180.go]: Nice solution, thanks. And removing the double endian decoding is much better code.

**Comment by fd0** [comment]
Ok, more improvements, thanks!

File: cmd/bmp180/main.go [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+	if interval != 0 {
+		t = time.NewTicker(interval)
+	}
+
+	for {
+		var env devices.Environment
+		if err := e.Sense(&env); err != nil {
+			return err
+		}
+		fmt.Printf("%8s %10s %9s\n", env.Temperature, env.Pressure, env.Humidity)
```
Comment (maruel) [inline cmd/bmp180/main.go]: There's no humidity sensing so it should not be printed.

File: experimental/devices/bmp180/bmp180.go [hunk 26]
```diff
@@ -0,0 +1,265 @@
...
+	O4x Oversampling = 2
+	O8x Oversampling = 3
+)
+
+// bit offsets in regCtrlMeas
+const (
+	ctrlMeasurementControlShift = 0
+	ctrlStartConversionShift    = 5
+	ctrlOversamplingShift       = 6
+)
```
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: optional nit: use one const () section, with one empty line being the separation between logical groups.

File: experimental/devices/bmp180/bmp180.go [hunk 27]
```diff
@@ -0,0 +1,265 @@
...
+func isValidU(i uint16) bool {
+	return i != 0 && i != 0xFFFF
+}
+
+// valid checks whether the calibration data is valid.
+func (c *calibration) isValid() bool {
+	return isValid(c.AC1) && isValid(c.AC2) && isValid(c.AC3) && isValidU(c.AC4) && isValidU(c.AC5) && isValidU(c.AC6) && isValid(c.B1) && isValid(c.B2) && isValid(c.MB) && isValid(c.MC) && isValid(c.MD)
+}
+
+// newCalibration parses calibration data from the buf and checks that it is valid.
```
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: remove now dead code

File: experimental/devices/bmp180/bmp180_test.go [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+		AC5: 32757,
+		AC6: 23153,
+		B1:  6190,
+		B2:  4,
+		MB:  -32768,
+		MC:  -8711,
+		MD:  2868,
+	}
+
+	temp := c.compensateTemp(27898)
```
Comment (maruel) [inline experimental/devices/bmp180/bmp180_test.go]: merge lines 24-25 into one
merge lines 39-40 into one

File: experimental/devices/bmp180/bmp180.go [hunk 29]
```diff
@@ -0,0 +1,265 @@
...
+	chipID      = 0x55 // contents of the ID register, always 0x55
+)
+
+// registers
+const (
+	regChipID           = 0xD0 // register contains the chip id
+	regCalibrationStart = 0xAA // first calibration register address
+	regSoftReset        = 0xE0 // soft reset register
+	regCtrlMeas         = 0xF4 // control measurement
+	regOutMSB           = 0xF6 // MSB of measurement
```
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: why not just `regOut = 0xF6 // 3 bytes register with measurement data`
the rest is not used directly anyway.

File: experimental/devices/bmp180/bmp180.go [hunk 30]
```diff
@@ -0,0 +1,265 @@
...
+	id, err := d.dev.ReadUint8(regChipID)
+	if err != nil {
+		return nil, err
+	}
+
+	if id != chipID {
+		return nil, fmt.Errorf("bmp180: unexpected chip id 0x%x; is this a BMP180?", id)
+	}
+
+	// issue soft reset to initialize device
```
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: I don't see why this is needed? I'm fine with exposing `Reset()` as a method but I don't see the need to call it every time the device is instantiated.

File: experimental/devices/bmp180/bmp180.go (hunk 26 above)
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: FTR, the reason is that I care about vertical density; the more dense the code can be vertically, the more a screenful page shows more code, leading to better visualization. Obviously this is a trade off with readability, as some people need more vertical spacing than others but I feel it is generally abused so I prefer not too many empty lines.

File: cmd/bmp180/main.go (hunk 25 above)
Comment (fd0) [inline cmd/bmp180/main.go]: OK

File: experimental/devices/bmp180/bmp180.go (hunk 26 above)
Comment (fd0) [inline experimental/devices/bmp180/bmp180.go]: Works for me.

File: experimental/devices/bmp180/bmp180.go (hunk 29 above)
Comment (fd0) [inline experimental/devices/bmp180/bmp180.go]: OK.

File: experimental/devices/bmp180/bmp180.go (hunk 30 above)
Comment (fd0) [inline experimental/devices/bmp180/bmp180.go]: You're right: It is not necessary (in most cases). I'd like to make sure that the device is in a known good state before interacting with it. We can move this to a `Reset()` method, though. What do you think?

**Comment by fd0** [comment]
Let me know if there's anything else.

File: experimental/devices/bmp180/bmp180.go (hunk 30 above)
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: Yes please. It's the last request after that I will merge the PR.

Comment (fd0) [inline experimental/devices/bmp180/bmp180.go]: done.

File: experimental/devices/bmp180/bmp180.go [hunk 31]
```diff
@@ -0,0 +1,247 @@
...
+var pressureConvTime = [...]time.Duration{
+	4500 * time.Microsecond,
+	7500 * time.Microsecond,
+	13500 * time.Microsecond,
+	25500 * time.Microsecond,
+}
+
+// Dev is a handle to a bmp180.
+type Dev struct {
+	dev *mmr.Dev8
```
Comment (maruel) [inline experimental/devices/bmp180/bmp180.go]: I just realized you don't need it to be a pointer, please inline instead of allocating on the heap. Last comment I promise.

Comment (fd0) [inline experimental/devices/bmp180/bmp180.go]: He, don't worry, when working with external contributors for [restic](https://github.com/restic/restic) (my project) I'm regularly on the other side of this.
